lib/Sema/TypeCheckType.cpp

```diff
@@ -646,25 +646,34 @@ static Type checkConstrainedExtensionRequirements(Type type,
     return type;
   }
 
-  // FIXME: Some day the type might also have its own 'where' clause, even
-  // if its not generic.
-
-  auto *ext = dyn_cast<ExtensionDecl>(decl->getDeclContext());
-  if (!ext || !ext->isConstrainedExtension())
-    return type;
-
-  if (parentTy->hasUnboundGenericType() ||
+  if (!parentTy || parentTy->hasUnboundGenericType() ||
       parentTy->hasTypeVariable()) {
     return type;
   }
 
-  auto subMap = parentTy->getContextSubstitutions(ext);
+  // We are interested in either a contextual where clause or
+  // a constrained extension context.
+  TypeSubstitutionMap subMap;
+  GenericSignature genericSig;
+  SourceLoc noteLoc;
+  if (decl->getTrailingWhereClause()) {
+    subMap = parentTy->getContextSubstitutions(decl->getDeclContext());
```

Comment:
I believe this and the next two lines should handle the extension case as well. Can you try collapsing down this 'if' statement in another PR?

Reply:
Am I missing something that is supposed to change in this branch if we are also inside a constrained extension? – the generic signature of the decl does the job of capturing the extension signature, if any.